packages/core/src/utils/GroupGraph.ts

```diff
@@ -0,0 +1,103 @@
+import { Shape, ShapeAD } from "../types/shape";
+import Graph from "./Graph";
+
+export type GroupGraphNode<T extends Shape | ShapeAD> = {
+  shape: T;
+  index: number;
+};
+
+export type GroupGraph<T extends Shape | ShapeAD> = Graph<
+  string,
+  GroupGraphNode<T>
+>;
+
+const makeGroupGraphNode = <T extends Shape | ShapeAD>(
+  shape: T,
+  index: number
+): [string, GroupGraphNode<T>, string[]] => {
+  const shapeType = shape.shapeType;
+  const shapeNameVal = shape.properties["name"];
+  if (shapeNameVal.tag !== "StrV") {
```

Comment:
Unrelated to this PR: this type of boilerplate should go away after we improve the shape types.